src/widget/wstarrating.cpp

```diff
@@ -166,7 +174,11 @@ void WStarRating::mouseReleaseEvent(QMouseEvent* /*unused*/) {
         return;
     }
 
-    m_pCurrentTrack->setRating(m_starRating.starCount());
+    m_currRating = m_starRating.starCount();
+    if (m_updateTrack) {
```

Comment:
Thanks for addressing these issues. QML will hopefully help to prevent those mistakes in the (far away) future.

The widget should never directly update the track state on its own. Instead the receiver of the signal could do that if desired or otherwise reject the proposed new value by resetting the widget -> unidirectional data flow.

Offering both options is confusing and not maintainable. But I am afraid that trying to avoid this mixed behavior could open another _Pandora's Box_.

Reply:
Ideally the widget would only offer a slot for setting the rating (input, write-only) and emit a signal when the rating changed (output, read-only). Maybe a getter for reading the current rating on demand. No track pointer, no implicit context management.

Reply:
I understand what you're saying, and I agree that WStarRating becomes more complex and bit harder to maintain. Looks like this can be changed without too much hazzle, with a few changes to Track and BaseTrackPlayer(Impl) (and mediators like LegacySkinParser, and DlgTrackinfo, of course), and WStarRating would be quite clean then.

I think I'll take a stab at this for 2.4 beta, though it's out of scope for this quick bug fix.